pkg/server/util/regentryutil/cache.go

```diff
@@ -0,0 +1,61 @@
+package regentryutil
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/spiffe/spire/proto/spire/common"
+)
+
+// FetchRegistrationEntriesCache is a wrapper around LRU cache with expiry, used for caching registration entries of a agent
+type FetchRegistrationEntriesCache struct {
+	Cache   *lru.Cache
```

Comment:
I think @azdagron had some thoughts on whether we should even use an LRU cache at all, or just a TTL-based cache, since ideally we'd want a cache size large enough that we never hit the limit, in which case we're not exactly relying on the LRU behavior. Maybe @azdagron  can comment a little more on that note?

Reply:
Considering the usage pattern for this API (every agent hitting it every 5 seconds), the LRU is a little weird. If the LRU is big enough to hold all entries, then we aren't really benefiting from an LRU and stale entries hang around indefinitely. If the LRU is too small for the number of entries, then large portions of it will be getting flushed quite a bit every 5 seconds, based on the ordering of the agent's hitting it.

I think a TTL based cache would probably be a better fit. It can still do proactive pruning on `Get` if an entry is expired and then periodically do a scan (performantly if the cache entries are ordered by expiration) to remove stale entries that haven't been used (in the case of an evicted or otherwise no-longer-used agent entry). I'd be interested in maybe some data (even with a toy implementation) that would demonstrate how effective the cache would be (e.g. # of hits/misses) given some common deployment modelling.

Reply:
Thanks @azdagron!
Let's take following example: 
```
cache ttl 1 sec, 
agent sync interval 5 sec,
workloads: 100, 
agents: 1000,
each agent has 10 workloads, (agent1: workload1..workload10, agent2: workload2..workload11).
500 agent sync between 0-1sec and next 500 between 1-2sec
```

Time between 0 to 1 sec:
```
agent1 invokes FetchX509SVID -> cache miss: all (cache will now contain: agent1, workload1..workload10)

agent2 invokes FetchX509SVID-> cache miss: agent2,worklaod11(2), cache hit: workload2..workload10(9).
```
...

In each sync interval, there will be 10k(1000 agents * 10 workloads) workload fetch calls, out of which there will be 500 cache misses(workload num * sync interval) in worst case.
In summary, each workload (and their children which are not considered in above example) will be fetched only once every second(ttl).

With respect to which cache to be used, as you correctly pointed out LRU is not helpful if we can accommodate all entries in cache.
But if that's not the case then LRU will persist those mapped registrations which are likely to be called again, over the node registrations which are likely to be not called again during cache ttl.

I'm wondering if we should assume that we can fit all entries in spire server memory given our continuing efforts to scale and gains we get by using just ttl based cache instead of LRU. 
I'd like to know your thoughts on this.

Reply:
We discussed this a bit during our maintainer sync. The consensus is that a sufficiently large LRU is probably OK for now until we understand more how this will impact memory usage and can make a more informed caching strategy decision. The large LRU will essentially work as a full cache for typical usage but would bound the cache from eating up too much memory in a burst of atypical registrations. 

Afar as how large the LRU should be in this first pass, I think we want something large, maybe 500K?

Reply:
sure..500K sounds good to me.